data_model/src/visit.rs

```diff
@@ -251,436 +274,188 @@ pub fn visit_wasm<V: Visit + ?Sized>(
+            Upgrade => visit_upgrade(),
+            Sequence => visit_sequence(),
+            Pair => visit_pair(),
         }
-        Expression::Where(expr) => visitor.visit_expression(authority, expr.expression()),
-        Expression::Query(query) => visitor.visit_query(authority, query),
-        Expression::ContextValue(_) | Expression::Raw(_) => {}
     }
 }
 
 pub fn visit_register<V: Visit + ?Sized>(
     visitor: &mut V,
     authority: &AccountId,
-    isi: &RegisterExpr,
+    isi: &RegisterBox,
 ) {
-    macro_rules! match_all {
-        ( $( $visitor:ident($object:ident) ),+ $(,)? ) => {
-            let object = evaluate_expr!(visitor, authority, <isi as Register>::object());
-
-            match object { $(
-                RegistrableBox::$object(object) => visitor.$visitor(authority, Register{object}), )+
-            }
-        };
-    }
-
     match_all! {
-        visit_register_peer(Peer),
-        visit_register_domain(Domain),
-        visit_register_account(Account),
-        visit_register_asset_definition(AssetDefinition),
-        visit_register_asset(Asset),
-        visit_register_role(Role),
-        visit_register_trigger(Trigger),
+        visitor, authority, isi,
+        RegisterBox {
+            Peer => visit_register_peer(),
+            Domain => visit_register_domain(),
+            Account => visit_register_account(),
+            AssetDefinition => visit_register_asset_definition(),
+            Asset => visit_register_asset(),
+            Role => visit_register_role(),
+            Trigger => visit_register_trigger(),
+        }
     }
 }
 
 pub fn visit_unregister<V: Visit + ?Sized>(
     visitor: &mut V,
     authority: &AccountId,
-    isi: &UnregisterExpr,
+    isi: &UnregisterBox,
 ) {
-    macro_rules! match_all {
-        ( $( $visitor:ident($id:ident) ),+ $(,)? ) => {
-            let object_id = evaluate_expr!(visitor, authority, <isi as Unregister>::object_id());
-            match object_id { $(
-                IdBox::$id(object_id) => visitor.$visitor(authority, Unregister{object_id}), )+
-                _ => visitor.visit_unsupported(authority, isi),
-            }
-        };
-    }
-
     match_all! {
-        visit_unregister_peer(PeerId),
-        visit_unregister_domain(DomainId),
-        visit_unregister_account(AccountId),
-        visit_unregister_asset_definition(AssetDefinitionId),
-        visit_unregister_asset(AssetId),
-        visit_unregister_role(RoleId),
-        visit_unregister_trigger(TriggerId),
+        visitor, authority, isi,
+        UnregisterBox {
+            Peer => visit_unregister_peer(),
+            Domain => visit_unregister_domain(),
+            Account => visit_unregister_account(),
+            AssetDefinition => visit_unregister_asset_definition(),
+            Asset => visit_unregister_asset(),
+            Role => visit_unregister_role(),
+            Trigger => visit_unregister_trigger(),
+        }
```

Comment:
I'm not sure this is much better than a plain `match` statement. There is no real code reduction, only obfuscation

Reply:
Agree, that's some overengineering